src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,372 @@
+    change quot.mk _ _ = _,
+    rw quot.sound rel.add_comm,
+    apply quot.sound,
+    apply rel.zero_add,
+  end,
+  add_comm := λ a b,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    apply quot.sound,
+    apply rel.add_comm,
+  end,
+  mul := λ a b, quot.lift_on a (λ x, quot.lift_on b (λ y, quot.mk _ (x * y))
+  begin
+    intros a b h,
+    dsimp only [],
+    apply quot.sound,
+    apply rel.mul_compat_right h,
+  end)
+  begin
+    intros a b h,
+    dsimp only [],
+    congr,
+    ext,
+    apply quot.sound,
+    apply rel.mul_compat_left h,
+  end,
+  mul_assoc := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.mul_assoc,
+  end,
+  one := quot.mk _ 1,
+  one_mul := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.one_mul,
+  end,
+  mul_one := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.mul_one,
+  end,
+  left_distrib := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.left_distrib,
+  end,
+  right_distrib := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.right_distrib,
+  end,
+  zero_mul := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.zero_mul,
+  end,
+  mul_zero := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.mul_zero,
+  end }
+
+instance : inhabited (tensor_algebra R M) := ⟨0⟩
+
+instance : has_scalar R (tensor_algebra R M) :=
+{ smul := λ r a, quot.lift_on a (λ x, quot.mk _ $ ↑r * x)
+begin
+  intros a b h,
+  dsimp only [],
+  apply quot.sound,
+  apply rel.mul_compat_right h,
+end }
+
+instance : algebra R (tensor_algebra R M) :=
+{ to_fun := λ r, quot.mk _ ↑r,
+  map_one' := rfl,
+  map_mul' := λ a b, by apply quot.sound; apply rel.mul_scalar,
+  map_zero' := rfl,
+  map_add' := λ a b, by apply quot.sound; apply rel.add_scalar,
+  commutes' := λ r x,
+  begin
+    dsimp only [],
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    apply quot.sound,
+    apply rel.central_scalar,
+  end,
+  smul_def' := λ r x, by refl }
+
+/--
+The universal linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def univ : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, by apply quot.sound; apply rel.add_lin,
+  map_smul' := λ r x, by apply quot.sound; apply rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
```

Comment:
Am I right in that the resulting g satisfies `g(coe(m1)*coe(m2)) = f(m1)*f(m2)`?

Reply:
Which map is g?

Reply:
If you're referring to `lift R M f` then the answer is yes -- it's an algebra map, and its composition with `univ R M` is exactly `f` by the theorem `univ_comp_lift`.

But note that there is no coercion introduced from `M` to `tensor_algebra R M` (but it's easy to define as `univ R M`, if needed).